course02/homework02/Tomasz/message_throttler_commons.hpp

```diff
@@ -0,0 +1,64 @@
+#pragma once
+
+#include <map>
+#include <chrono>
+
+#include "short_circular_buffer.hpp"
+#include "message_throttler.hpp"
+#include "message_throttler_interface.hpp"
+
+template<typename _Message>
+struct message_swallower
+{
+	void operator()(const _Message& message)
+	{ }
+};
+
+struct chrono_timestamp_threshold
+{
+	chrono_timestamp_threshold(std::chrono::milliseconds milliseconds) :
+		milliseconds(milliseconds)
+	{ }
+
+	bool operator()(std::chrono::milliseconds now, std::chrono::milliseconds timestamp)
+	{
+		return (now - timestamp) < milliseconds;
+	}
+
+	std::chrono::milliseconds milliseconds;
+};
+
+struct chrono_timestamper
+{
+	std::chrono::milliseconds operator()()
+	{
+		return std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::system_clock::now().time_since_epoch());
+	}
+};
+
+template<
+	typename _ClientId,
+	std::size_t _BufferSize,
+	typename _Message,
+	typename _Timestamp = std::chrono::milliseconds,
+	typename _MessageConsumer = message_swallower<_Message>,
+	typename _MessageDisposer = message_swallower<_Message>,
+	typename _Timestamper = chrono_timestamper,
+	typename _TimestampThreshold = chrono_timestamp_threshold,
+	typename _Buffer = short_circular_buffer<_Timestamp, _BufferSize>,
+	typename _MessageThrottlerInterface = message_throttler_interface<_Message, _MessageConsumer, _MessageDisposer, _Timestamp, _Timestamper, _TimestampThreshold, _Buffer>,
+	typename _Map = std::map<_ClientId, std::unique_ptr<_MessageThrottlerInterface>>
```

Comment:
It's nice to template for customization points, although I must admit that you took that a bit on the extreme side :) Typically, the map isn't really needed: If you have one class SlidingWindow<Message, ...> with most of the logic, the outside class should be dead simple - just an unordered_map of client to SlidingWindow.  Which is why it doesnt need to be templated - the client code can just be changed. Timestamp and Timestamper can be replaced if you template on the Clock type - then you know your time point/duration/etc types and the related operations. BufferSize and TimestampThreshold should be parameters, not compile time constants imho.

Reply:
I've changed all of the _Timestampe[Threshold|er] parameters with just a single _Clock parameter. And I also removed _Map from parameters.